content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md

```diff
@@ -184,6 +182,14 @@ for `kube-apiserver`, `kube-controller-manager`, `kube-scheduler` and `etcd` to
 If the flag is not set, those components run as root. You can change the value of this feature gate before
 you upgrade to a newer version of Kubernetes.
 
+`WaitForAllControlPlaneComponents`
```

Comment:
When does this wait starts and ends?
Why should I care about this gate?

Reply:
> When does this wait starts and ends?

NewWaitControlPlanePhase is a hidden phase that runs after the control-plane and etcd phases. It is a phase in `kubeadm init` and `kubeadm join` for a control-plane node.

It is after `etcd`&`control-plane` and before `upload-config`.

In `kubeadm init`,  this hidden phase is after `NewEtcdPhase` and `NewControlPlanePhase`.
- before this FG, we only wait for API server ready.
- with this FG enabled, we will wait for all control-plane components including KCM and Scheduler.
https://github.com/kubernetes/kubernetes/blob/87412809b7df44da06c5fa54f8b29d37012693af/cmd/kubeadm/app/cmd/init.go#L156-L158

In `kubeadm join`, it is the last phase after general control-plane joins(in the joining, etcd is ready and node is marked as control-plane and )
https://github.com/kubernetes/kubernetes/blob/87412809b7df44da06c5fa54f8b29d37012693af/cmd/kubeadm/app/cmd/join.go#L222-L224

> Why should I care about this gate?

We add the FG to keep the old behavior without enabling it. In v1.30, it will be tested in alpha level. Later we will shift to beta and GA. 

If users want to wait for all components to be ready before init/join control-plan exits, they can enable their FG now.  

https://github.com/kubernetes/kubeadm/issues/2907 is a use case.

Reply:
this explains it well, @pacoxu thanks.
i personally think the level of detail in the pr diff is enough for the users, but happy to amend something.

Reply:
adding the sentence:
> The wait process starts right after the kubelet on the host is started by kubeadm